osm2streets/src/lanes/placement.rs

```diff
@@ -0,0 +1,170 @@
+use abstutil::Tags;
+
+use super::{LtrLaneNum, Placement, RoadPosition};
+
+use Placement::*;
+use RoadPosition::*;
+
+impl TryFrom<&str> for RoadPosition {
+    type Error = ();
+
+    /// Tries to parse a road position from an osm tag value as per the `placement` scheme.
+    /// See https://wiki.openstreetmap.org/wiki/Proposed_features/placement#Tagging
+    ///
+    /// The direction is treated as forward, use `reverse()` on the result if the context is backward.
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        match value {
+            "" => Ok(Center),
+            "separation" => Ok(Separation),
+            _ => {
+                if let Some((kind, lane_str)) = value.split_once(':') {
+                    if let Ok(lane) = lane_str.parse::<usize>() {
+                        match kind {
+                            "left_of" => Ok(LeftOf(LtrLaneNum::Forward(lane))),
+                            "middle_of" => Ok(MiddleOf(LtrLaneNum::Forward(lane))),
+                            "right_of" => Ok(RightOf(LtrLaneNum::Forward(lane))),
+                            _ => Err(()),
+                        }
+                    } else {
+                        Err(())
+                    }
+                } else {
+                    Err(())
+                }
+            }
+        }
+    }
+}
+
+impl TryFrom<&Tags> for Placement {
```

Comment:
I am very happy with treating the parsing step of the whole scheme independently from seeing if the semantics of the tag values match the road they are tagged on.

Do you think I should back out from implementing the traits though? Or is this the sort of thing you are interested in working towards?

Reply:
The separate parsing is awesome, and the unit tests show the niceness. I would vote to just change the API to `impl Placement { fn parse(tags: &Tags) -> anyhow::Result<Self> }` though. I'm not sure how the wider Rust community feels or what's standard in codebases, but the from/into traits don't show up as clearly in rustdocs and they're visually less clear to spot in code. (Something like `let x: Placement = some_complicated_thing().try_into()?` vs `let x = Placement::parse(some_complicated_thing())?`)

Reply:
Yep, `anyhow` or any other error type lets us use `?`. `anyhow` is encouraged for applications and discouraged for libraries, but similar to the thread below about error/warning handling, I don't think increased code complexity is worth it here yet. We can be lazy and use string errors